src/mpi/comm/comm_split_type.c

```diff
@@ -91,14 +91,88 @@ int MPIR_Comm_split_type_self(MPIR_Comm * comm_ptr, int key, MPIR_Comm ** newcom
     goto fn_exit;
 }
 
-int MPIR_Comm_split_type_hw_guided(MPIR_Comm * user_comm_ptr, int key,
+static const char *info_get_key(MPIR_Info * info_ptr, const char *key)
+{
+    MPIR_Info *curr_ptr = info_ptr->next;
+    while (curr_ptr) {
+        if (strcmp(curr_ptr->key, key) == 0) {
+            return curr_ptr->value;
+        }
+        curr_ptr = curr_ptr->next;
+    }
+    return NULL;
+}
+
+int MPIR_Comm_split_type_hw_guided(MPIR_Comm * comm_ptr, int key,
                                    MPIR_Info * info_ptr, MPIR_Comm ** newcomm_ptr)
 {
     int mpi_errno = MPI_SUCCESS;
+    MPIR_Comm *node_comm = NULL;
+    const char *resource_type = NULL;
 
-    *newcomm_ptr = NULL;
+    if (info_ptr != NULL) {
+        resource_type = info_get_key(info_ptr, "mpi_hw_resource_type");
+    }
+
+    if (!resource_type) {
+        /* we need compare the info hints regardless since it is collective */
+        resource_type = "";
+    }
+
+    /* Check all processes are using the same hints, return NULL if not.
+     * Note: since it is a collective checking, we need call this function
+     * even when info_ptr is NULL or the key is missing. */
+    int is_equal = 0;
+    mpi_errno = MPII_compare_info_hint(resource_type, comm_ptr, &is_equal);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    MPIR_ERR_CHKANDJUMP1(!is_equal, mpi_errno, MPI_ERR_OTHER, "**infonoteq",
+                         "**infonoteq %s", "mpi_hw_resource_type");
+
+    if (resource_type[0] == '\0') {
+        /* return MPI_COMM_NULL if info key is missing */
+        *newcomm_ptr = NULL;
+        goto fn_exit;
+    }
+
+    if (strcmp(resource_type, "mpi_shared_memory") == 0) {
+        /* resource_type value "mpi_shared_memory" is equivalent to split_type
+         * MPI_COMM_TYPE_SHARED */
+        mpi_errno = MPIR_Comm_split_type_impl(comm_ptr, MPI_COMM_TYPE_SHARED, key, info_ptr,
+                                              newcomm_ptr);
+        MPIR_ERR_CHECK(mpi_errno);
+        goto fn_exit;
+    }
 
+    /* now we should proceed */
+    mpi_errno = MPIR_Comm_split_type_by_node(comm_ptr, key, &node_comm);
+    MPIR_ERR_CHECK(mpi_errno);
+
+    if (!MPIR_hwtopo_is_initialized()) {
+        /* if hwtopo is not available, just return the node_comm */
+        *newcomm_ptr = node_comm;
+        node_comm = NULL;
+        goto fn_exit;
+    }
```

Comment:
>     * Would you mind describing the diffs between `MPIR_Comm_split_type_hw_guided ` and current `shmem_topo ` in the commit message? I feel that it will help followers understand the changed logic

(I'll address other two points separately ). I think now that we have official HW_GUIDED, the `shmem_topo` hints feature should get deleted. I am planning to do so right after we merge this PR. That's why I don't think we need too much comments on `shmem_topo`. What do you think?

Reply:
>  I think now that we have official HW_GUIDED, the shmem_topo hints feature should get deleted. I am planning to do so right after we merge this PR. That's why I don't think we need too much comments on shmem_topo. What do you think?

Sounds good to me